test/app-luatest/trigger_module_test.lua

```diff
@@ -0,0 +1,498 @@
+    trigger.del(event, tostring(break_point))
+    local deleted_id = math.random(break_point + 1, 100)
+    trigger.del(event, tostring(deleted_id))
+
+    for _, handler in unpack(iter) do
+        handler()
+    end
+
+    t.assert_equals(#history, trigger_num - 1)
+    local history_idx = 0
+    for i = 1, trigger_num do
+        if i ~= deleted_id then
+            history_idx = history_idx + 1
+            t.assert_equals(history[history_idx], i)
+            trigger.del(event, tostring(i))
+        end
+    end
+end
+
+-- Check that pairs is successfully stopped after all triggers registered
+-- on the event are dropped
+g.test_clear_event_during_pairs = function()
+    local trigger = require('trigger')
+
+    local event = 'event'
+    local trigger_num = 100
+    local break_point = math.random(5, 95)
+
+    local history = {}
+    local function handler_method(self)
+        table.insert(history, self.id)
+    end
+    local mt = {__call = handler_method}
+    for i = trigger_num, 1, -1 do
+        local a = {id = i}
+        setmetatable(a, mt)
+        trigger.set(event, tostring(i), a)
+    end
+
+    local idx = 0
+    local iter = {trigger.pairs(event)}
+    for _, handler in unpack(iter) do
+        idx = idx + 1
+        handler()
+        if idx == break_point then
+            break
+        end
+    end
+
+    for i = 1, trigger_num do
+        trigger.del(event, tostring(i))
+    end
+
+    for _, handler in unpack(iter) do
+        handler()
+    end
+
+    t.assert_equals(#history, break_point)
+    for i = 1, break_point do
+        t.assert_equals(history[i], i)
+    end
+end
+
+g.test_invalid_args = function()
+    local trigger = require('trigger')
+
+    local valid_name = "name"
+
+    local invalid_handlers = {nil, 42, 'abc', {}, {1, 2}, {k='v'}, box.NULL}
+    local errmsg = 'callable expected'
+    for _, h in pairs(invalid_handlers) do
+        t.assert_error_msg_contains(errmsg, trigger.set, valid_name,
+            valid_name, h)
+    end
+
+    local handler = function() return true end
+    local invalid_names = {nil, {}, {1, 2}, {k='v'}, box.NULL, handler}
+    errmsg = 'string expected'
+    for _, invalid_name in pairs(invalid_names) do
+        t.assert_error_msg_contains(
+            errmsg, trigger.set, valid_name, invalid_name, handler)
+        t.assert_error_msg_contains(
+            errmsg, trigger.set, invalid_name, valid_name, handler)
+        t.assert_error_msg_contains(
+            errmsg, trigger.del, valid_name, invalid_name)
+        t.assert_error_msg_contains(
+            errmsg, trigger.del, invalid_name, valid_name)
+        t.assert_error_msg_contains(errmsg, trigger.call, invalid_name)
+        t.assert_error_msg_contains(errmsg, trigger.pairs, invalid_name)
+        t.assert_error_msg_contains(errmsg, trigger.info, invalid_name)
+    end
+    errmsg = "expected .* argument.*"
+    t.assert_error_msg_matches(errmsg, trigger.set, 'event', 'trigger',
+        handler, 'redundant')
```

Comment:
Please check all error messages properly with `assert_error_msg_equals` or `assert_error_msg_content_equals`, not with `assert_error_msg_matches` or `assert_error_msg_contains`.

Reply:
Done.